File: src/lib/desktop/payroll.ts
import { invoke } from '@tauri-apps/api/core';
import type { TauriActionResult } from '@/lib/desktop/admin-types';
import { isTauriBackendAvailable } from '@/lib/desktop/runtime';

function requireTauri(): void {
  if (!isTauriBackendAvailable()) {
    throw new Error('Tauri payroll API requires the desktop shell');
  }
}

export async function tauriPayrollGetOrgMembersForHr() {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('payroll_get_org_members_for_hr_cmd');
}

export async function tauriPayrollCreateRecord(data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('payroll_create_record_cmd', { data });
}

export async function tauriPayrollGetRecord(recordId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('payroll_get_record_cmd', { recordId });
}

export async function tauriPayrollGetRecordsByOrg(orgId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('payroll_get_records_by_org_cmd', { orgId });
}

export async function tauriPayrollGetRecordsForCurrentOrg(limit?: number) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('payroll_get_records_for_current_org_cmd', { limit: limit ?? null });
}

export async function tauriPayrollGetRecordsByEmployee(employeeId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('payroll_get_records_by_employee_cmd', { employeeId });
}

export async function tauriPayrollGetRecordsByPeriod(orgId: string, payPeriod: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('payroll_get_records_by_period_cmd', { orgId, payPeriod });
}

export async function tauriPayrollUpdateRecord(recordId: string, data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('payroll_update_record_cmd', { recordId, data });
}

export async function tauriPayrollDeleteRecord(recordId: string) {
  requireTauri();
  return invoke<TauriActionResult<boolean>>('payroll_delete_record_cmd', { recordId });
}

export async function tauriPayrollCreateLeaveRequest(data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('payroll_create_leave_request_cmd', { data });
}

export async function tauriPayrollGetLeaveRequest(requestId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('payroll_get_leave_request_cmd', { requestId });
}

export async function tauriPayrollGetLeaveRequestsByOrg(orgId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('payroll_get_leave_requests_by_org_cmd', { orgId });
}

export async function tauriPayrollGetLeaveRequestsForCurrentOrg(limit?: number) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('payroll_get_leave_requests_for_current_org_cmd', { limit: limit ?? null });
}

export async function tauriPayrollGetLeaveRequestsByEmployee(employeeId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('payroll_get_leave_requests_by_employee_cmd', { employeeId });
}

export async function tauriPayrollGetLeaveRequestsByStatus(orgId: string, status: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('payroll_get_leave_requests_by_status_cmd', { orgId, status });
}

export async function tauriPayrollUpdateLeaveRequest(requestId: string, data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('payroll_update_leave_request_cmd', { requestId, data });
}

export async function tauriPayrollDeleteLeaveRequest(requestId: string) {
  requireTauri();
  return invoke<TauriActionResult<boolean>>('payroll_delete_leave_request_cmd', { requestId });
}

File: src-tauri/src/commands/payroll.rs
use serde_json::Value;
use tauri::State;

use crate::admin::service::ActionResult;
use crate::auth::session::SessionState;
use crate::payroll::leave_service::{
    create_leave_request, delete_leave_request, get_leave_request, get_leave_requests_by_employee,
    get_leave_requests_by_org, get_leave_requests_by_status, get_leave_requests_for_current_org,
    update_leave_request,
};
use crate::payroll::service::{
    create_payroll_record, delete_payroll_record, get_org_members_for_hr, get_payroll_record,
    get_payroll_records_by_employee, get_payroll_records_by_org, get_payroll_records_by_period,
    get_payroll_records_for_current_org, update_payroll_record, HrEmployee,
};
use crate::state::AppState;

#[tauri::command]
pub fn payroll_get_org_members_for_hr_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
) -> ActionResult<Vec<HrEmployee>> {
    get_org_members_for_hr(&app, &session)
}

#[tauri::command]
pub fn payroll_create_record_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    data: Value,
) -> ActionResult<Value> {
    create_payroll_record(&app, &session, data)
}

#[tauri::command]
pub fn payroll_get_record_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    record_id: String,
) -> ActionResult<Value> {
    get_payroll_record(&app, &session, &record_id)
}

#[tauri::command]
pub fn payroll_get_records_by_org_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: String,
) -> ActionResult<Vec<Value>> {
    get_payroll_records_by_org(&app, &session, &org_id)
}

#[tauri::command]
pub fn payroll_get_records_for_current_org_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    limit: Option<usize>,
) -> ActionResult<Vec<Value>> {
    get_payroll_records_for_current_org(&app, &session, limit.unwrap_or(500))
}

#[tauri::command]
pub fn payroll_get_records_by_employee_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    employee_id: String,
) -> ActionResult<Vec<Value>> {
    get_payroll_records_by_employee(&app, &session, &employee_id)
}

#[tauri::command]
pub fn payroll_get_records_by_period_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: String,
    pay_period: String,
) -> ActionResult<Vec<Value>> {
    get_payroll_records_by_period(&app, &session, &org_id, &pay_period)
}

#[tauri::command]
pub fn payroll_update_record_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    record_id: String,
    data: Value,
) -> ActionResult<Value> {
    update_payroll_record(&app, &session, &record_id, data)
}

#[tauri::command]
pub fn payroll_delete_record_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    record_id: String,
) -> ActionResult<bool> {
    delete_payroll_record(&app, &session, &record_id)
}

#[tauri::command]
pub fn payroll_create_leave_request_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    data: Value,
) -> ActionResult<Value> {
    create_leave_request(&app, &session, data)
}

#[tauri::command]
pub fn payroll_get_leave_request_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    request_id: String,
) -> ActionResult<Value> {
    get_leave_request(&app, &session, &request_id)
}

#[tauri::command]
pub fn payroll_get_leave_requests_by_org_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: String,
) -> ActionResult<Vec<Value>> {
    get_leave_requests_by_org(&app, &session, &org_id)
}

#[tauri::command]
pub fn payroll_get_leave_requests_for_current_org_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    limit: Option<usize>,
) -> ActionResult<Vec<Value>> {
    get_leave_requests_for_current_org(&app, &session, limit.unwrap_or(300))
}

#[tauri::command]
pub fn payroll_get_leave_requests_by_employee_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    employee_id: String,
) -> ActionResult<Vec<Value>> {
    get_leave_requests_by_employee(&app, &session, &employee_id)
}

#[tauri::command]
pub fn payroll_get_leave_requests_by_status_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: String,
    status: String,
) -> ActionResult<Vec<Value>> {
    get_leave_requests_by_status(&app, &session, &org_id, &status)
}

#[tauri::command]
pub fn payroll_update_leave_request_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    request_id: String,
    data: Value,
) -> ActionResult<Value> {
    update_leave_request(&app, &session, &request_id, data)
}

#[tauri::command]
pub fn payroll_delete_leave_request_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    request_id: String,
) -> ActionResult<bool> {
    delete_leave_request(&app, &session, &request_id)
}
